internal/services/dispatch_test.go

```diff
@@ -190,6 +189,107 @@ func TestDispatchIntegration(t *testing.T) {
 				require.Error(t, cerr)
 			},
 		},
+		{
+			"delete preconditions test",
+			`definition user {}
+
+			definition resource {
+				relation viewer: user
+				permission view = viewer
+			}`,
+			func(t *testing.T, client v1.PermissionsServiceClient) {
+				// Ensure the delete fails on the precondition.
+				_, derr := client.DeleteRelationships(context.Background(), &v1.DeleteRelationshipsRequest{
+					OptionalPreconditions: []*v1.Precondition{
+						{
+							Operation: v1.Precondition_OPERATION_MUST_MATCH,
+							Filter: &v1.RelationshipFilter{
+								ResourceType:       "resource",
+								OptionalResourceId: "someresource",
+								OptionalRelation:   "viewer",
+								OptionalSubjectFilter: &v1.SubjectFilter{
+									SubjectType:       "user",
+									OptionalSubjectId: "sarah",
+								},
+							},
+						},
+					},
+					RelationshipFilter: &v1.RelationshipFilter{
+						ResourceType: "resource",
+					},
+				})
+				require.Error(t, derr)
+				require.Contains(t, derr.Error(), "unable to satisfy write precondition")
+
+				// Write a relationship, but not the one we want.
+				_, werr := client.WriteRelationships(context.Background(), &v1.WriteRelationshipsRequest{
+					Updates: []*v1.RelationshipUpdate{
+						{
+							Operation:    v1.RelationshipUpdate_OPERATION_CREATE,
+							Relationship: tuple.MustToRelationship(tuple.MustParse("resource:someresource#viewer@user:someuser")),
+						},
+					},
+				})
+				require.NoError(t, werr)
+
+				// Ensure the delete still fails on the precondition.
+				_, derr = client.DeleteRelationships(context.Background(), &v1.DeleteRelationshipsRequest{
+					OptionalPreconditions: []*v1.Precondition{
+						{
+							Operation: v1.Precondition_OPERATION_MUST_MATCH,
+							Filter: &v1.RelationshipFilter{
+								ResourceType:       "resource",
+								OptionalResourceId: "someresource",
+								OptionalRelation:   "viewer",
+								OptionalSubjectFilter: &v1.SubjectFilter{
+									SubjectType:       "user",
+									OptionalSubjectId: "sarah",
+								},
+							},
+						},
+					},
+					RelationshipFilter: &v1.RelationshipFilter{
+						ResourceType: "resource",
+					},
+				})
+				require.Error(t, derr)
+				require.Contains(t, derr.Error(), "unable to satisfy write precondition")
+
+				// Write the relationship needed.
+				_, werr = client.WriteRelationships(context.Background(), &v1.WriteRelationshipsRequest{
+					Updates: []*v1.RelationshipUpdate{
+						{
+							Operation:    v1.RelationshipUpdate_OPERATION_CREATE,
+							Relationship: tuple.MustToRelationship(tuple.MustParse("resource:someresource#viewer@user:sarah")),
+						},
+					},
+				})
+				require.NoError(t, werr)
+
+				// Ensure the delete now works.
```

Comment:
Can you add a `MUST_NOT_MATCH` after the write to verify that that also fails now.

Reply:
Done